### PR TITLE
Store MCP Server Secrets in Settings

### PR DESCRIPTION
**Pull request by aiyakitori**
## What does this PR do?

Headers and Env Vars previous were stored as part of the data model which can be synced. This PR adds the ability for tool servers to save their "secrets" such as API Key inside Kiln Settings. 

* Add logic to handle secrets lifecycle in ExternalToolServer.
* Update _Tool Server_ screen to mask secrets and indicate if a key is missing simulating initial sync state. 
* Add status column to  _Manage Tool_ screen to indicate if a tool server has missing keys. 
* Update connect tool server screen to mark if a header or env var is a secret. 

## Checklists

- [x] Tests have been run locally and passed
- [x] New tests have been added to any work in /lib

image
image

## Summary by CodeRabbit

- **New Features**
  - Centralized MCP secret storage (MCP_SECRETS_KEY); tool-server secrets tracked separately, masked in UI, and listings expose missing_secrets with Status (Ready / Action Required).

- **Refactor**
  - Unified key/value model for headers/env vars with per-item "Secret" toggles in add-tool UIs. Creation payloads include secret_header_keys/secret_env_var_keys. Sessions merge secrets into requests/env without mutating originals. Shared UiProperty type introduced for property lists.

- **Tests**
  - Expanded tests for validation, secret lifecycle, session merging, UI flows, and edge cases.

### CODE REVIEW

File: app/desktop/studio_server/tool_api.py [hunk 1]
```diff
@@ -87,6 +89,24 @@ def validate_server_details(self):
             # Update headers to validated version
             self.headers = validated_headers
 
+        # Validate secret header keys
+        if isinstance(self.secret_header_keys, list):
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: little strange to check the type here. Pydantic will check the type automatically so it's not needed. Doesn't hurt but not needed.

File: app/desktop/studio_server/tool_api.py [hunk 2]
```diff
@@ -87,6 +89,24 @@ def validate_server_details(self):
...
+            validated_secret_header_keys = []
+            for key in self.secret_header_keys:
+                key_str = key.strip() if isinstance(key, str) else str(key).strip()
+                if not key_str:
+                    raise ValueError("Secret header key is required")
+                # Check if the key is in the headers
+                if key_str not in self.headers:
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: validate key is in, not key_str (stripped). You'll get into weird error states where the value isn't actually valid: `    key` will pass validation but really shouldn't.

Striping is good to check empty, but don't use it here.

File: app/desktop/studio_server/tool_api.py [hunk 3]
```diff
@@ -87,6 +89,24 @@ def validate_server_details(self):
...
+                    raise ValueError(
+                        f"Secret header key {key_str} is not in the headers"
+                    )
+
+                validated_secret_header_keys.append(key_str)
+
+            # Update secret header keys to validated version
+            self.secret_header_keys = validated_secret_header_keys
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: hard rule: no mutations in validators. 

Just check it's valid above.

In general: kill "magic". We could spend an hour debugging why something sometimes works, only to find the validator is magically fixing a bad value. Just error on the bad values and we'll catch the bug in 10 seconds.

Comment (scosman) [reply app/desktop/studio_server/tool_api.py]: If you wanted to check "no leading/trailing whitespace" then check that directly, and raise a nice error explaining the specific issue so use can fix. But never just hide the issue. P3 to do that though.

File: app/desktop/studio_server/tool_api.py [hunk 4]
```diff
@@ -120,6 +141,24 @@ def validate_command(self):
...
+                # Check if the key is in the env_vars
+                if key_str not in self.env_vars:
+                    raise ValueError(
+                        f"Secret environment variable key {key_str} is not in the list of environment variables"
+                    )
+
+                validated_secret_env_var_keys.append(key_str)
+
+            # Update secret env var keys to validated version
+            self.secret_env_var_keys = validated_secret_env_var_keys
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: ditto to all of above

File: app/desktop/studio_server/tool_api.py [hunk 5]
```diff
@@ -304,6 +345,27 @@ async def get_tool_server(
...
+            created_by=tool_server.created_by,
+            properties=tool_server.properties,
+            available_tools=[],
+            missing_secrets=[],
+        )
+
+        # Check if the tool server has missing secretes (e.g. new user syncing exisiting project)
+        # If there are missing secrets, add a requirement to the result and skip getting available tools.
+        missing_secrets = tool_server.missing_secrets()
+        if missing_secrets and len(missing_secrets) > 0:
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: len check and other checks unnecessary. If it's an array of zero that's valid. Just assign directly (assuming method doesn't return None, but just make it return empty array and not None if it does).

File: app/desktop/studio_server/tool_api.py [hunk 6]
```diff
@@ -304,6 +345,27 @@ async def get_tool_server(
...
+        # Create a default result with incomplete status
+        result = ExternalToolServerApiDescription(
+            id=tool_server.id,
+            name=tool_server.name,
+            type=tool_server.type,
+            description=tool_server.description,
+            created_at=tool_server.created_at,
+            created_by=tool_server.created_by,
+            properties=tool_server.properties,
+            available_tools=[],
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: nit: I like the old way more. Having a partly instantiated but incorrect object could introduce buggy. Just calculate values first and build the object when you have it (like you had)

This pattern is sometimes needed with dicts, but not objects.

File: app/desktop/studio_server/tool_api.py [hunk 7]
```diff
@@ -320,32 +382,27 @@ async def get_tool_server(
...
         properties = {
             "server_url": tool_data.server_url,
             "headers": tool_data.headers,
+            "secret_header_keys": tool_data.secret_header_keys,
         }
 
         tool_server = ExternalToolServer(
-            name=tool_data.name,
-            type=ToolServerType.remote_mcp,  # Default to remote MCP type
+            name=tool_data.name.strip(),  # Trim whitespace for filename validation
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: no magic edits. Let it error on invalid inputs.

File: app/desktop/studio_server/tool_api.py [hunk 8]
```diff
@@ -365,15 +425,16 @@ async def connect_local_mcp(
...
         properties = {
             "command": tool_data.command,
             "args": tool_data.args,
             "env_vars": tool_data.env_vars,
+            "secret_env_var_keys": tool_data.secret_env_var_keys,
         }
 
         tool_server = ExternalToolServer(
-            name=tool_data.name,
+            name=tool_data.name.strip(),  # Trim whitespace for filename validation
```
Comment (scosman) [inline app/desktop/studio_server/tool_api.py]: no magic

File: app/web_ui/src/lib/ui/property_list.svelte [hunk 9]
```diff
@@ -24,6 +25,8 @@
     <div class="flex items-center text-gray-500 overflow-x-hidden">
```
Comment (scosman) [inline app/web_ui/src/lib/ui/property_list.svelte]: put error up here `<div class="flex items-center overflow-x-hidden {property.error ? 'text-error' : 'text-gray-500'}">`

that way you can have errors and links, and it's not either-or.

File: app/web_ui/src/lib/ui/property_list.svelte [hunk 10]
```diff
@@ -24,6 +25,8 @@
...
       {#if property.link}
         <a href={property.link} class="link">{property.value}</a>
+      {:else if property.error}
```
Comment (scosman) [inline app/web_ui/src/lib/ui/property_list.svelte]: remove per above

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/+page.svelte [hunk 11]
```diff
@@ -155,6 +171,9 @@
                     Disable
                   </button>
                 </td>
+                <td class="text-sm">
+                  <span class="text-success">✓ Ready</span>
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/+page.svelte]: Use same same code so it remains visually consistent: 

```
 <Warning
                      warning_message="Ready"
                      warning_color="success"
                      warning_icon="check"
                    />
```

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte [hunk 12]
```diff
@@ -7,21 +7,28 @@
...
     server_url: string
-    headers: { key: string; value: string; placeholder: string | null }[]
+    headers: KeyValuePair[]
   }
 
-  interface LocalMcpServer {
-    name: string
-    subtitle: string
-    description: string
+  interface LocalMcpServer extends BaseMcpServer {
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte]: of fancy. If you do move to types, syntax will change but same thing possible.

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte [hunk 13]
```diff
@@ -74,11 +82,13 @@
           key: "Authorization",
           value: "apikey REPLACE_WITH_TWELVE_DATA_API_KEY",
           placeholder: "apikey REPLACE_WITH_TWELVE_DATA_API_KEY",
+          is_secret: true,
         },
         {
           key: "X-OpenAPI-Key",
           value: "",
           placeholder: "REPLACE_WITH_YOUR_OPENAI_API_KEY",
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte:85]: I swear we already addressed this in a PR, but these value placeholder strings need work.

Placeholders don't need to be all caps, shouldn't have "REPLACE_". Can just use nice plain English "Your Firecrawl API Key", "Your OpenAI API Key". The REPLACE is ugly, we only used it for cases where we anted to prepopulate the value because it needs a special format (e.g. "Bearer ...."). I think all placeholders need work.

Again: I swear I did this.... is there another PR pending?

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte [hunk 14]
```diff
@@ -7,21 +7,28 @@
 
   $: project_id = $page.params.project_id
 
-  interface RemoteMcpServer {
+  interface KeyValuePair {
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte]: not sure why you're using interface instead of type. I think here you have  more typical types? I'm not really sure here though, not an expert. This might be better.

Are TS interfaces more for real interfaces/protocols (things external objects implement)?

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte [hunk 15]
```diff
@@ -11,17 +11,18 @@
   import Warning from "$lib/ui/warning.svelte"
 
   // Environment Variables as array of key/value pairs
-  interface EnvVarPair {
+  interface KeyValuePair {
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte]: repeated code several places. 

You should define a types using a .ts file. It can live in /src/lib/... then import it in all 3 places

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte [hunk 16]
```diff
@@ -54,23 +55,37 @@
...
+    envVarsObj: Record<string, string>
+    secret_env_var_keys: string[]
+  } {
     const envVarsObj: Record<string, string> = {}
+    const secretEnvVarKeys: string[] = []
 
     for (const envVar of env_vars) {
       if (envVar.key.trim() && envVar.value.trim()) {
-        envVarsObj[envVar.key.trim()] = envVar.value.trim()
+        const key = envVar.key.trim()
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte]: no magic

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte [hunk 17]
```diff
@@ -54,23 +55,37 @@
...
-  function buildEnvVarsObject(): Record<string, string> {
+  function buildEnvVarsObject(): {
+    envVarsObj: Record<string, string>
+    secret_env_var_keys: string[]
+  } {
     const envVarsObj: Record<string, string> = {}
+    const secretEnvVarKeys: string[] = []
 
     for (const envVar of env_vars) {
       if (envVar.key.trim() && envVar.value.trim()) {
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte:60]: okay here since we're ignoring completely blank rows (good UI design in UI layer)

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte [hunk 18]
```diff
@@ -54,23 +55,37 @@
...
-        envVarsObj[envVar.key.trim()] = envVar.value.trim()
+        const key = envVar.key.trim()
+        envVarsObj[key] = envVar.value.trim()
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte]: no magic (especially since spaces are valid values)

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte [hunk 19]
```diff
@@ -208,6 +225,20 @@
...
           </div>
+          <div class="flex-1 max-w-[100px]">
+            <FormElement
+              inputType="select"
+              label="Secret"
+              id="secret_{item_index}"
+              info_description="If this environment variable is a secret such as an API key, select 'Yes' to prevent it from being synced. Kiln will store the secret in your project's settings."
+              light_label={true}
+              select_options={[
+                [false, "No"],
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte]: "Secret" and "Not secret" (and remove "select 'yes'" above)

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte [hunk 20]
```diff
@@ -15,13 +15,14 @@
   let description = ""
 
   // Headers as array of key/value pairs
-  interface HeaderPair {
+  interface KeyValuePair {
     key: string
     value: string
     placeholder: string | null
+    is_secret: boolean
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte]: ditto

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte [hunk 21]
```diff
@@ -46,24 +47,36 @@
...
+    headersObj: Record<string, string>
+    secret_header_keys: string[]
+  } {
     const headersObj: Record<string, string> = {}
+    const secretHeaderKeys: string[] = []
 
     for (const header of headers) {
       if (header.key.trim() && header.value.trim()) {
-        headersObj[header.key.trim()] = header.value.trim()
+        const key = header.key.trim()
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte]: no magic

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte [hunk 22]
```diff
@@ -46,24 +47,36 @@
...
+        headersObj[key] = header.value.trim()
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte]: no magic (and headers can have spaces)

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte [hunk 23]
```diff
@@ -179,6 +194,20 @@
...
           </div>
+          <div class="flex-1 max-w-[100px]">
+            <FormElement
+              inputType="select"
+              label="Secret"
+              id="secret_{item_index}"
+              info_description="If this header is a secret such as an API key, select 'Yes' to prevent it from being synced. Kiln will store the secret in your project's settings."
+              light_label={true}
+              select_options={[
+                [false, "No"],
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte]: ditto

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte [hunk 24]
```diff
@@ -134,6 +134,67 @@
     return properties
   }
 
+  type UiProperty = {
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte]: don't re-define types. This already exists. (I see this mistake is made elsewhere, but don't copy that)

1) Two options: move it into exported type in a .ts file next to property_list.svelte, import it both here and in property_list.svelte
2) Just let typescript do its thing. Don't define the return value. Typescript should dynamically infer it, and give you warnings if the data is incorrect. Check this but should work.

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte [hunk 25]
```diff
@@ -134,6 +134,67 @@
...
+    const properties: UiProperty[] = []
+    properties.push(
+      ...Object.entries(noneSecretProperties).map(([key, value]) => ({
+        name: key,
+        value: String(value ?? "N/A"),
+      })),
+    )
+
+    // Make sure secretKeys is an array of strings
+    const secretKeysArray: string[] = Array.isArray(secretKeys)
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte]: don't dynamically type check unless needed -- the method specifies an array of strings, so it will be. Type checker will catch callers passing the wrong thing. Keep the code as typed as possible. Use dynamic type checking to make it typed asap, but not constantly validating after it's typed.

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte [hunk 26]
```diff
@@ -134,6 +134,67 @@
...
+
+  /**
+   * Helper function to build properties list with secret handling
+   * @param noneSecretProperties - Object containing non-secret key-value pairs
+   * @param secretKeys - Array of secret key names
+   * @param missingSecrets - Array of missing secret key names
+   * @returns Array of UiProperty objects
+   */
+  function buildPropertiesWithSecrets(
+    noneSecretProperties: Record<string, string>,
```
Comment (scosman) [inline app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte]: nit: nonSecretProperties, not none

File: libs/core/kiln_ai/datamodel/external_tool_server.py [hunk 27]
```diff
@@ -62,6 +63,16 @@ def validate_required_fields(self) -> "ExternalToolServer":
                     raise ValueError(
                         "headers must be a dictionary for external tools of type 'remote_mcp'"
                     )
+
+                secret_header_keys = self.properties.get("secret_header_keys", None)
+                # Secret header keys are optional, but if they are set, they must be a list of strings
```
Comment (scosman) [inline libs/core/kiln_ai/datamodel/external_tool_server.py:101]: you're checking list, but not strings

File: libs/core/kiln_ai/datamodel/external_tool_server.py [hunk 28]
```diff
@@ -83,7 +94,182 @@ def validate_required_fields(self) -> "ExternalToolServer":
                         "environment variables must be a dictionary for external tools of type 'local_mcp'"
                     )
 
+                secret_env_var_keys = self.properties.get("secret_env_var_keys", None)
```
Comment (scosman) [inline libs/core/kiln_ai/datamodel/external_tool_server.py:131]: ditto

File: libs/core/kiln_ai/datamodel/external_tool_server.py [hunk 29]
```diff
@@ -83,7 +94,182 @@ def validate_required_fields(self) -> "ExternalToolServer":
...
+            # Look for secrets with the pattern: mcp_server_id::key_name
+            if mcp_secrets:  # Only proceed if mcp_secrets is not None
+                for key_name in secret_keys:
+                    secret_value = mcp_secrets.get(self._config_secret_key(key_name))
+                    if secret_value:
+                        secrets[key_name] = secret_value
+
+        return secrets
+
+    def missing_secrets(self) -> list[str]:
```
Comment (scosman) [inline libs/core/kiln_ai/datamodel/external_tool_server.py]: nit: almost same method as above. Combine, and do tuple return Tuple[secrets, missing_secrets]

File: libs/core/kiln_ai/datamodel/external_tool_server.py [hunk 30]
```diff
@@ -83,7 +94,182 @@ def validate_required_fields(self) -> "ExternalToolServer":
...
+        secret_keys = self.get_secret_keys()
+
+        if not secret_keys:
+            return
+
+        if self.id is None:
+            raise ValueError("Server ID cannot be None when saving secrets")
+
+        # Extract secret values from properties based on server type
+        match self.type:
```
Comment (scosman) [inline libs/core/kiln_ai/datamodel/external_tool_server.py]: use `get_secret_keys`

File: libs/core/kiln_ai/datamodel/external_tool_server.py [hunk 31]
```diff
@@ -83,7 +94,182 @@ def validate_required_fields(self) -> "ExternalToolServer":
...
+            case ToolServerType.remote_mcp:
+                secret_values = self.properties.get("headers", {})
+            case ToolServerType.local_mcp:
+                secret_values = self.properties.get("env_vars", {})
+            case _:
+                raise_exhaustive_enum_error(self.type)
+
+        config = Config.shared()
+        mcp_secrets = config.get_value(MCP_SECRETS_KEY) or dict[str, str]()
```
Comment (scosman) [inline libs/core/kiln_ai/datamodel/external_tool_server.py]: use explicit type definition on var instead of the `dict[str, str]()` to give it implicit type

`mcp_secrets: dict[str, str] = config.get_value(MCP_SECRETS_KEY) or {}`

File: libs/core/kiln_ai/datamodel/external_tool_server.py [hunk 32]
```diff
@@ -83,7 +94,182 @@ def validate_required_fields(self) -> "ExternalToolServer":
...
+        # Remove secrets with the pattern: mcp_server_id::key_name
+        for key_name in secret_keys:
+            secret_key = self._config_secret_key(key_name)
+            if secret_key in mcp_secrets:
+                del mcp_secrets[secret_key]
+
+        # Always call update_settings to maintain consistency with the old behavior
+        config.update_settings({MCP_SECRETS_KEY: mcp_secrets})
+
+    def save_to_file(self) -> None:
```
Comment (scosman) [inline libs/core/kiln_ai/datamodel/external_tool_server.py:252]: You end up having 2 inconsistent object states for the same object: before save secrets are in headers, after they aren't. Callers could easily have bugs.

The object also doesn't work unless functions are called in the right order (if you call save to file before save_secrets, the secrets are already deleted).

Suggestion: Process everything in `__init__` instead. Convert to the correct representation right away (no secrets in header/env_vars, just keys in secret_header_keys/secret_env_var_keys). You can store the secrets in memory in an internal non-model variable by setting `this._unsaved_secrets = {...}` in init. Still override save_to_file, but just to check if _unsaved_secrets isn't empty, and then call  internal _save_secrets (remove the public save_secrets). That way you can't forget and just save file, or secrets, but not both. Might need to update retrieve_secrets to merge config secrets and local `this._unsaved_secrets`

File: app/desktop/studio_server/tool_api.py (hunk 5 above)
Comment (aiyakitori) [reply app/desktop/studio_server/tool_api.py]:
```
def missing_secrets(self) -> list[str]:
```
This returns empty array guaranteed.

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte (hunk 12 above)
Comment (aiyakitori) [reply app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte]: used types for consistency

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte (hunk 24 above)
Comment (aiyakitori) [reply app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/tool_servers/[tool_server_id]/+page.svelte]: Did 1, i had this initially.

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte (hunk 13 above)
Comment (aiyakitori) [reply app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/+page.svelte:85]: Didn't see any pending PRs around this. We can open a small PR to fix the strings?

File: app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte [hunk 33]
```diff
@@ -208,6 +218,20 @@
...
+              id="secret_{item_index}"
+              info_description="If this environment variable is a secret such as an API key, select 'Secret' to prevent it from being synced. Kiln will store the secret in your project's settings."
+              light_label={true}
+              select_options={[
+                [false, "No Secret"],
+                [true, "Secret"],
+              ]}
+              bind:value={env_vars[item_index].is_secret}
+            />
+          </div>
```
[the remark this answers is not in the document]
Comment (aiyakitori) [reply app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/local_mcp/+page.svelte:237]: We explicitly want a dropdown so it looks better.

**Review by scosman — APPROVED** [review]
